File: frontend/src/models/referralStateMachine.ts
import type { ReferralStateT } from "./careEpisode";

/**
 * This table is a MIRROR of the backend's authoritative transition rules.
 * It exists so the UI only ever presents legal next-actions to the user
 * (better UX, fewer wasted taps) — but the backend re-validates every
 * transition independently. Never trust this file alone for correctness;
 * see Build Guide Section 3 and Section 11 (server-side enforcement).
 */
export const LEGAL_TRANSITIONS: Record<ReferralStateT, ReferralStateT[]> = {
  DRAFT: ["SENT", "CANCELLED"],
  SENT: ["RECEIVED", "ACCEPTED", "REJECTED", "EXPIRED", "EMERGENCY_ESCALATED"],
  RECEIVED: ["ACCEPTED", "REJECTED", "EMERGENCY_ESCALATED"],
  ACCEPTED: ["APPOINTMENT_QUEUED", "CONSULTED", "PATIENT_NO_SHOW", "EMERGENCY_ESCALATED"],
  APPOINTMENT_QUEUED: ["CONSULTED", "PATIENT_NO_SHOW", "EMERGENCY_ESCALATED"],
  CONSULTED: ["REFERRED_BACK", "FOLLOW_UP_DUE", "CLOSED"],
  REFERRED_BACK: ["FOLLOW_UP_DUE", "CLOSED"],
  FOLLOW_UP_DUE: ["FOLLOW_UP_COMPLETED", "PATIENT_NO_SHOW"],
  FOLLOW_UP_COMPLETED: ["CLOSED"],
  CLOSED: [],
  REJECTED: [],
  CANCELLED: [],
  EXPIRED: [],
  PATIENT_NO_SHOW: ["SENT"], // allow re-referral after a no-show
  EMERGENCY_ESCALATED: ["ACCEPTED", "CLOSED"],
};

export function isLegalTransition(
  from: ReferralStateT,
  to: ReferralStateT
): boolean {
  return LEGAL_TRANSITIONS[from]?.includes(to) ?? false;
}

export function getAvailableTransitions(
  current: ReferralStateT
): ReferralStateT[] {
  return LEGAL_TRANSITIONS[current] ?? [];
}

export const TERMINAL_STATES: ReferralStateT[] = [
  "CLOSED",
  "REJECTED",
  "CANCELLED",
  "EXPIRED",
];

export function isTerminal(state: ReferralStateT): boolean {
  return TERMINAL_STATES.includes(state);
}

/** Human-readable labels for the UI — keep display concerns out of the model. */
export const STATE_LABELS: Record<ReferralStateT, string> = {
  DRAFT: "Draft",
  SENT: "Sent",
  RECEIVED: "Received",
  ACCEPTED: "Accepted",
  APPOINTMENT_QUEUED: "Appointment Queued",
  CONSULTED: "Consulted",
  REFERRED_BACK: "Referred Back",
  FOLLOW_UP_DUE: "Follow-up Due",
  FOLLOW_UP_COMPLETED: "Follow-up Completed",
  CLOSED: "Closed",
  REJECTED: "Rejected",
  CANCELLED: "Cancelled",
  EXPIRED: "Expired",
  PATIENT_NO_SHOW: "No-Show",
  EMERGENCY_ESCALATED: "Emergency Escalated",
};

File: backend/app/services/referral_state_machine.py
from app.models.referral import ReferralState, TERMINAL_STATES

"""
This is the AUTHORITATIVE version of the transition table. The frontend
has a mirror (src/models/referralStateMachine.ts) purely for UX — so the
UI only ever offers legal next-actions — but this file is what actually
gets enforced. Never trust client input alone for a state transition
(Build Guide Section 11: "enforced state machine, not a free-text status
field").

Keep this table byte-for-byte in sync with the frontend's copy. If they
diverge, the frontend might offer an action the backend then rejects —
annoying but safe (fails closed). The reverse (backend more permissive
than frontend) is the dangerous direction, so if in doubt, keep this
table the stricter of the two.
"""

LEGAL_TRANSITIONS: dict[ReferralState, set[ReferralState]] = {
    ReferralState.DRAFT: {ReferralState.SENT, ReferralState.CANCELLED},
    ReferralState.SENT: {
        ReferralState.RECEIVED,
        ReferralState.ACCEPTED,
        ReferralState.REJECTED,
        ReferralState.EXPIRED,
        ReferralState.EMERGENCY_ESCALATED,
    },
    ReferralState.RECEIVED: {
        ReferralState.ACCEPTED,
        ReferralState.REJECTED,
        ReferralState.EMERGENCY_ESCALATED,
    },
    ReferralState.ACCEPTED: {
        ReferralState.APPOINTMENT_QUEUED,
        ReferralState.CONSULTED,
        ReferralState.PATIENT_NO_SHOW,
        ReferralState.EMERGENCY_ESCALATED,
    },
    ReferralState.APPOINTMENT_QUEUED: {
        ReferralState.CONSULTED,
        ReferralState.PATIENT_NO_SHOW,
        ReferralState.EMERGENCY_ESCALATED,
    },
    ReferralState.CONSULTED: {
        ReferralState.REFERRED_BACK,
        ReferralState.FOLLOW_UP_DUE,
        ReferralState.CLOSED,
    },
    ReferralState.REFERRED_BACK: {ReferralState.FOLLOW_UP_DUE, ReferralState.CLOSED},
    ReferralState.FOLLOW_UP_DUE: {
        ReferralState.FOLLOW_UP_COMPLETED,
        ReferralState.PATIENT_NO_SHOW,
    },
    ReferralState.FOLLOW_UP_COMPLETED: {ReferralState.CLOSED},
    ReferralState.CLOSED: set(),
    ReferralState.REJECTED: set(),
    ReferralState.CANCELLED: set(),
    ReferralState.EXPIRED: set(),
    ReferralState.PATIENT_NO_SHOW: {ReferralState.SENT},  # allow re-referral
    ReferralState.EMERGENCY_ESCALATED: {ReferralState.ACCEPTED, ReferralState.CLOSED},
}


class IllegalTransitionError(Exception):
    def __init__(self, from_state: ReferralState, to_state: ReferralState):
        self.from_state = from_state
        self.to_state = to_state
        super().__init__(
            f"Illegal referral transition: {from_state.value} -> {to_state.value}"
        )


def validate_transition(from_state: ReferralState, to_state: ReferralState) -> None:
    """Raises IllegalTransitionError if the transition isn't allowed.
    Callers (the referrals router) catch this and return HTTP 409, not
    a generic 500 — this is a business-rule violation, not a bug."""
    allowed = LEGAL_TRANSITIONS.get(from_state, set())
    if to_state not in allowed:
        raise IllegalTransitionError(from_state, to_state)


def is_terminal(state: ReferralState) -> bool:
    return state in TERMINAL_STATES
